File: 1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        map = {}
        for (index, value) in enumerate(nums):
            # calculate the match needed to hit the target 
            diff = target - value

            # check if the match is in the map
            if diff in map:
                return [map[diff], index]
            
            # no match in map. Add value so match can find this value
            map[value] = index

        return [0,0]
    
        orderedNums = {}
        for i in range(len(nums)):
            if nums[i] in orderedNums:
                if nums[i] * 2 == target:
                    return [orderedNums[nums[i]], i]
            orderedNums[nums[i]] = i

        nums.sort()

        for i in range(len(nums)-1):
            for j in range(i+1, len(nums)):
                # rest are too big skip
                a = nums[i] + nums[j]
                if a > target:
                    break
                elif a == target:
                    return [orderedNums[nums[i]], orderedNums[nums[j]]]

        # shouldnt hit if there is a solution
        return []
        
# @lc code=end

File: rust/src/bin/1.two-sum.rs
/*
 * @lc app=leetcode id=1 lang=rust
 *
 * [1] Two Sum
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map: HashMap<i32, i32> = HashMap::new();
        
        for (index, &value) in nums.iter().enumerate() {
            // calculate the value of a pair
            let diff = target - value;

            // check hashmap for value - if it exists return the pair of indicies
            if let Some(pair) = map.get(&diff) {
                return vec![*pair, index as i32]; 
            }
            // no match found so lets add the corresponding pair to the map.
            map.insert(value, index as i32);
        }

        // shouldnt happen.
        return vec![0,0]
    }
}
// @lc code=end

